src/corehost/cli/coreclr.cpp

```diff
@@ -5,95 +5,132 @@
     const char** argv,
     const char* managedAssemblyPath,
     unsigned int* exitCode);
 
-static coreclr_shutdown_fn coreclr_shutdown = nullptr;
-static coreclr_initialize_fn coreclr_initialize = nullptr;
-static coreclr_execute_assembly_fn coreclr_execute_assembly = nullptr;
+// Prototype of the coreclr_create_delegate function from coreclr.dll
+using coreclr_create_delegate_fn = pal::hresult_t(STDMETHODCALLTYPE *)(
+    coreclr_t::host_handle_t hostHandle,
+    unsigned int domainId,
+    const char* entryPointAssemblyName,
+    const char* entryPointTypeName,
+    const char* entryPointMethodName,
+    void** delegate);
 
-bool coreclr::bind(const pal::string_t& libcoreclr_path)
+namespace
 {
-    assert(g_coreclr == nullptr);
-
-    pal::string_t coreclr_dll_path(libcoreclr_path);
-    append_path(&coreclr_dll_path, LIBCORECLR_NAME);
+    pal::dll_t g_coreclr = nullptr;
+    coreclr_shutdown_fn coreclr_shutdown = nullptr;
+    coreclr_initialize_fn coreclr_initialize = nullptr;
+    coreclr_execute_assembly_fn coreclr_execute_assembly = nullptr;
+    coreclr_create_delegate_fn coreclr_create_delegate = nullptr;
 
-    if (!pal::load_library(&coreclr_dll_path, &g_coreclr))
+    bool coreclr_bind(const pal::string_t& libcoreclr_path)
     {
-        return false;
-    }
+        assert(g_coreclr == nullptr);
 
-    coreclr_initialize = (coreclr_initialize_fn)pal::get_symbol(g_coreclr, "coreclr_initialize");
-    coreclr_shutdown = (coreclr_shutdown_fn)pal::get_symbol(g_coreclr, "coreclr_shutdown_2");
-    coreclr_execute_assembly = (coreclr_execute_assembly_fn)pal::get_symbol(g_coreclr, "coreclr_execute_assembly");
+        pal::string_t coreclr_dll_path(libcoreclr_path);
+        append_path(&coreclr_dll_path, LIBCORECLR_NAME);
 
-    return true;
-}
+        if (!pal::load_library(&coreclr_dll_path, &g_coreclr))
+        {
+            return false;
+        }
 
-void coreclr::unload()
-{
-    assert(g_coreclr != nullptr && coreclr_initialize != nullptr);
+        coreclr_initialize = (coreclr_initialize_fn)pal::get_symbol(g_coreclr, "coreclr_initialize");
+        coreclr_shutdown = (coreclr_shutdown_fn)pal::get_symbol(g_coreclr, "coreclr_shutdown_2");
+        coreclr_execute_assembly = (coreclr_execute_assembly_fn)pal::get_symbol(g_coreclr, "coreclr_execute_assembly");
+        coreclr_create_delegate = (coreclr_create_delegate_fn)pal::get_symbol(g_coreclr, "coreclr_create_delegate");
+
+        assert(coreclr_initialize != nullptr
+            && coreclr_shutdown != nullptr
+            && coreclr_execute_assembly != nullptr
+            && coreclr_create_delegate != nullptr);
 
-    pal::unload_library(g_coreclr);
+        return true;
+    }
+    
+    void coreclr_unload()
+    {
+        assert(g_coreclr != nullptr && coreclr_initialize != nullptr);
+
+        pal::unload_library(g_coreclr);
+    }
 }
 
-pal::hresult_t coreclr::initialize(
+pal::hresult_t coreclr_t::create(
+    const pal::string_t& libcoreclr_path,
```

Comment:
NIT: Any reason to change the name from initialize to create?  The delegate remains coreclr_initialize

Reply:
Because this is creating a `coreclr_t` instance. Initialization is an implementation detail of the type and is an implementation detail. This is a static factory method for the new `coreclr_t` type where as the previous implementation was designed around everything being static.